optuna/pruners/hyperband.py

```diff
@@ -0,0 +1,152 @@
+            is computed with this parameter and the index of bracket.
+        min_early_stopping_rate_high:
+            The maximum value of the minimum early stopping rate for `SuccessiveHalvingPruner`.
+            The number of brackets is
+            :math:`\\mathsf{min_early_stopping_rate_high} - \\mathsf{min_early_stopping_rate_low}`.
+    """
+
+    def __init__(
+            self,
+            min_resource=1,
+            reduction_factor=3,
+            min_early_stopping_rate_low=0,
+            min_early_stopping_rate_high=4
+    ):
+        # type: (int, int, int, int) -> None
+        self._pruners = []  # type: List[SuccessiveHalvingPruner]
+        self._reduction_factor = reduction_factor
+        self._resource_badget = 0
+        n_pruners = min_early_stopping_rate_high - min_early_stopping_rate_low + 1
+        self._n_pruners = n_pruners
+        self._bracket_resource_budgets = []  # type: List[int]
+
+        _logger.debug('Hypberband has {} brackets'.format(self.n_pruners))
+
+        for i in range(n_pruners):
+            bracket_resource_budget = self._calc_bracket_resource_budget(i, n_pruners)
+            self._resource_badget += bracket_resource_budget
+            self._bracket_resource_budgets.append(bracket_resource_budget)
+
+            # N.B. (crcrpar): `min_early_stopping_rate` has the information of `bracket_index`.
+            min_early_stopping_rate = min_early_stopping_rate_low + i
+            rung_key_prefix = 'bracket{}_'.format(i)
+
+            _logger.debug(
+                '{}th bracket has minimum early stopping rate of {}'.format(
+                    i, min_early_stopping_rate))
+
+            pruner = SuccessiveHalvingPruner(
+                min_resource=min_resource,
+                reduction_factor=reduction_factor,
+                min_early_stopping_rate=min_early_stopping_rate,
+                rung_key_prefix=rung_key_prefix
+            )
+            self._pruners.append(pruner)
+
+    def prune(self, study, trial):
+        # type: (Study, FrozenTrial) -> bool
+
+        i = self.get_bracket_id(study.study_id, trial.number)
+        _logger.debug('{}th bracket is selected'.format(i))
+        return self._pruners[i].prune(study, trial)
+
+    def get_trial_pruner_auxiliary_data(self, study_id, trial_number):
+        # type: (int, int) -> str
+
+        bracket_index = self.get_bracket_id(study_id, trial_number)
+        return '_{}'.format(bracket_index)
+
+    def should_filter_trials(self):
+        # type: () -> bool
+
+        return True
+
+    def _calc_bracket_resource_budget(self, pruner_index, n_pruners):
+        # type: (int, int) -> int
+        n = self._reduction_factor ** (n_pruners - 1)
+        budget = n
+        for i in range(pruner_index, n_pruners - 1):
+            budget += n / 2
+        return budget
+
+    def get_bracket_id(self, study_id, trial_number):
```

Comment:
As mentioned above, `study_id` is deprecated.
```suggestion
    def get_bracket_id(self, study_name, trial_number):
```